### ISSUE TITLE
runtimevar/etcdvar: etcd type mismatch

### ISSUE THREAD

**Issue by Lscchina** [google/go-cloud#1912]
### Describe the bug

I imported gocloud for version: gocloud.dev v0.13.0, and then when I compile my project , I get an error: cannot use cli (type *"go.etcd.io/etcd/clientv3".Client) as type *"github.com/coreos/etcd/clientv3".Client in argument to concurrency.NewSession

### To Reproduce
1.My project imported etcd for version: v0.0.0-20181109213058-9454c4cab09b
2.I  import gocloud for version: gocloud.dev v0.13.0, and use runtimevar package
3.Build my project, got an error: cannot use cli (type *"go.etcd.io/etcd/clientv3".Client) as type *"github.com/coreos/etcd/clientv3".Client in field value

## Expected behavior

use etcd version with package go.etcd.io/etcd, not github.com/coreos/etcd

### Version

1.gocloud version: v0.13.0
2.my local go version: 1.11.1
3.my local project use etcd version: go.etcd.io/etcd v0.0.0-20181109213058-9454c4cab09b

### Additional context

1.gocloud use go.etcd.io/etcd v3.3.12+incompatible
2.then I search in https://github.com/etcd-io/etcd/tree/v3.3.12, there is no go.mod, and in the project's go files, the package is github.com/coreos/etcd

**Comment by eliben** [comment]
Can you please clarify what you'd like to see happen here?

AFAICT, the github dependency comes from within etcd itself. Consider:

```
$ go mod why -m go.etcd.io/etcd
# go.etcd.io/etcd
gocloud.dev/runtimevar/etcdvar
go.etcd.io/etcd/clientv3
```

And:

```
$ go mod why -m github.com/coreos/etcd
# github.com/coreos/etcd
gocloud.dev/runtimevar/etcdvar
go.etcd.io/etcd/clientv3
github.com/coreos/etcd/auth/authpb
```

And in our `go.mod` file it's marked as indirect for this reason.

Can you provide a small reproducer that demonstrates your problem? If I undestand correctly you have a project that doesn't build because of etcd version mismatch

**Comment by Lscchina** [comment]
import (
	"context"
	"fmt"
	"math/rand"
	"time"
	
	"go.etcd.io/etcd/clientv3"
	"go.etcd.io/etcd/clientv3/concurrency"
	"gocloud.dev/runtimevar"
)

// KeepAlive 注册服务并保活，srv服务名称，addr可访问的地址，ttl单位是秒
func KeepAlive(ctx context.Context, cli *clientv3.Client, srv, addr string, ttl int) (err error) {
	for {
		err = keepAlive(ctx, cli, srv, addr, ttl)
		if err == context.DeadlineExceeded || err == context.Canceled {
			return
		}
		time.Sleep(time.Duration(rand.Int63n(10)) * time.Millisecond)
	}
}

// keepAlive 注册服务并保活
func keepAlive(ctx context.Context, cli *clientv3.Client, srv, addr string, ttl int) error {
	sess, err := concurrency.NewSession(cli, concurrency.WithContext(ctx), concurrency.WithTTL(ttl))
	if err != nil {
		return err
	}
	defer sess.Close()

	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-sess.Done():
		return fmt.Errorf("%s/%s exit", srv, addr)
	}
}

func TestGocloud() (*runtimevar.Variable) {
	return nil
}



------------------------------------------------------------------------------------
above is a test code, compile it will get the error

**Comment by eliben** [comment]
Thanks for the reproducer. If I remove all mentions of gocloud from it, I still get the same error. See this directory: https://github.com/eliben/gocloud-samples/tree/master/runtimevar-etcd

If you clone it and cd inside, and run `go build`, you get:

```
# sample
./sample.go:24:37: cannot use cli (type *"go.etcd.io/etcd/clientv3".Client) as type *"github.com/coreos/etcd/clientv3".Client in argument to concurrency.NewSession
```

**Comment by Lscchina** [comment]
Thanks. in your sample's go.mod file, the etcd version is "go.etcd.io/etcd v3.3.12+incompatible", it can cause the error. and this version is same with gocloud's etcd version.
If you use another version, such as "go.etcd.io/etcd v0.0.0-20181109213058-9454c4cab09b", it will run well.
```go
require go.etcd.io/etcd v0.0.0-20181109213058-9454c4cab09b
```
above is my go.mod

**Comment by eliben** [comment]
It seems like you should open an issue for the etcd repository. Please link it here so we can follow along. Their repo seems to have a go.mod, so I'm not 100% clear on why they released 3.3.12 without it

**Comment by eliben** [comment]
FWIW #1987 reduced some dependencies on etcd-modules, though it's not clear whether this will resoleve your problem. I'd still recommend opening an upstream issue for etcd

**Comment by Lscchina** [comment]
Thank you, i will open an issue for etcd

**Comment by Lscchina** [comment]
link from etcd https://github.com/etcd-io/etcd/issues/10480, they will fix this problom in v3.4

**Comment by vangent** [comment]
I don't think there's anything for us to do here; we will upgrade to 3.4 when it is available.
